File: src/modules/document/repository.py
"""문서 저장소 인터페이스."""
from abc import ABC, abstractmethod
from typing import Optional

from modules.document.model import Document


class DocumentRepository(ABC):
    """
    문서 저장소의 인터페이스.

    저장소는 문서를 저장하고 검색하는 메서드를 정의한다.
    구체적인 저장소 구현(메모리, 데이터베이스 등)은
    이 인터페이스를 구현해야 한다.
    """

    @abstractmethod
    def create(self, document: Document) -> Document:
        """
        새로운 문서를 저장소에 저장한다.

        Args:
            document: 저장할 문서

        Returns:
            저장된 문서

        Raises:
            다양한 저장소 구현별 예외가 발생할 수 있음
        """
        pass

    @abstractmethod
    def get(self, id: str) -> Optional[Document]:
        """
        주어진 id로 문서를 조회한다.

        Args:
            id: 조회할 문서의 고유 식별자

        Returns:
            조회된 문서 또는 없으면 None
        """
        pass

    @abstractmethod
    def get_by_normalized_title(self, normalized_title: str) -> Optional[Document]:
        """
        정규화된 제목으로 문서를 조회한다.

        Args:
            normalized_title: 조회할 문서의 정규화된 제목

        Returns:
            조회된 문서 또는 없으면 None
        """
        pass

File: php/src/Modules/Document/Repository.php
<?php

declare(strict_types=1);

namespace MintWiki\Document;

/**
 * 문서 저장소 포트 (태스크 0402).
 *
 * Python `DocumentRepository`(src/modules/document/repository.py)와 메서드
 * 이름/입력/출력/예외 계약을 맞춘 인터페이스. `docs/repository-port-contracts.md`
 * 의 document 섹션이 정본이다. 구현 없이 시그니처만 고정하며 — 이 포트를
 * 구현하는 InMemory/Database 구현체는 이후 태스크(0435 등)에서 추가된다.
 *
 * 클래스 이름은 `docs/php-namespace-mapping.md`가 고정한 규칙대로 이미
 * `MintWiki\Document` namespace 안에 있으므로 중복되는 `Document` 접두어를
 * 뺀다 (Python `DocumentService` -> PHP `MintWiki\Document\Service`와 동일한
 * 패턴).
 */
interface Repository
{
    /**
     * @throws DuplicateNormalizedTitleError normalizedTitle이 이미 존재하는 경우
     */
    public function create(Document $document): Document;

    public function get(string $id): ?Document;

    public function getByNormalizedTitle(string $normalizedTitle): ?Document;

    /**
     * @throws NotFoundError document의 id가 저장소에 없는 경우
     */
    public function update(Document $document): Document;
}
